### PR TITLE
Let the command line `--help` text be translatable

### LINKED ISSUE #3051: The DOSBox Staging help screen should be translatable
Originally posted by @Kappa971 in https://github.com/dosbox-staging/dosbox-staging/issues/884#issuecomment-1777376888

### PR DESCRIPTION
**Pull request by kcgen**
# Description

Adds the main executable CLI help message to the messaging system.

Excellent suggestion, @Kappa971.

## Related issues

Fixes #3051 

# Manual testing

Tested with CLI `--help`.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [x] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by kcgen** [comment]
@Kappa971, does this work on your side w/ an Italian translation?

File: src/gui/sdlmain.cpp [hunk 1]
```diff
@@ -4135,6 +4135,68 @@ static void messages_add_sdl()
...
+	        "\n"
+	        "  --exit                   Exit after the DOS program specified by FILE has ended.\n"
+	        "                           If no FILE has been specified, execute [autoexec] and\n"
+	        "                           terminate.\n"
+	        "\n"
+	        "  --startmapper            Run the mapper GUI.\n"
+	        "\n"
+	        "  --erasemapper            Delete the default mapper file.\n"
+	        "\n"
+	        "  --securemode             Enable secure mode. The [config] and [autoexec] sections\n"
```
Comment (FeralChild64) [inline src/gui/sdlmain.cpp:4191]: Do we have a `[config]` section at all?

**Comment by Kappa971** [comment]
> @Kappa971, does this work on your side w/ an Italian translation?

I compiled this build but I can't see the new strings.

**Comment by Kappa971** [comment]
`  -c <command>             Run the specified DOS command before running FILE.`

What is FILE? 😅

File: src/gui/sdlmain.cpp [hunk 2]
```diff
@@ -4135,6 +4135,68 @@ static void messages_add_sdl()
 	        "seamless mouse, %s+F10 to capture");
 	MSG_Add("TITLEBAR_HINT_SEAMLESS_HOTKEY_MIDDLE",
 	        "seamless mouse, %s+F10 or middle-click to capture");
+
+	MSG_Add("PROGRAM_DOSBOX_HELP_LONG",
```
[the remark this answers is not in the document]
Comment (kcgen) [reply src/gui/sdlmain.cpp]: Ok :+1:

**Comment by kcgen** [comment]
> > @Kappa971, does this work on your side w/ an Italian translation?
> 
> I compiled this build but I can't see the new strings.

Update inbound. See manual testing procedure in the PR comment 

> ` -c <command> Run the specified DOS command before running FILE.`
> 
> What is FILE? 😅

Update inbound. Added to the `DOSBOX_HELP` message. Snippet:

```c++
"FILE     A mountable path such as a directory, drive image\n"
"         (IMG/ISO/CUE), or DOS executable (BAT/COM/EXE). While\n"
"         all paths will be mounted, executables are additionally\n"
"         run, unmounted, and terminated.\n");
```

File: src/gui/sdlmain.cpp (hunk 1 above)
Comment (kcgen) [reply src/gui/sdlmain.cpp:4173]: Good catch; will snip this errant reference.

Comment (kcgen) [reply src/gui/sdlmain.cpp]: Removed.

Comment (FeralChild64) [reply src/gui/sdlmain.cpp]: `[autoexec]` is still mentioned in `--securemode` description.

Comment (kcgen) [reply src/gui/sdlmain.cpp]: removed; thanks for re-mentioning [autoexec], @FeralChild64.

**Comment by johnnovak** [comment]
> > > @Kappa971, does this work on your side w/ an Italian translation?
> > 
> > 
> > I compiled this build but I can't see the new strings.
> 
> Update inbound.
> 
> > ` -c <command> Run the specified DOS command before running FILE.`
> > What is FILE? 😅
> 
> Added to the `DOSBOX_HELP` message. Snippet:
> 
> ```c++
> "FILE     A mountable path such as a directory, drive image\n"
> "         (IMG/ISO/CUE), or DOS executable (BAT/COM/EXE). While\n"
> "         all paths will be mounted, executables are additionally\n"
> "         run, unmounted, and terminated.\n");
> ```

"unmounted, and terminated" - how is an executable unmounted? Terminated; well, the executable terminates at some point unless it contains an infinite loop 😄 

Can the wording be more precise and less confusing? E.g., just plainly state that paths are mounted, and for executable args their containing folder is mounted as the C drive and the executable is run.

File: src/gui/sdlmain.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnnovak) [reply src/gui/sdlmain.cpp]: Agreed, I expect the `PROGRAM_` prefix/scope for DOS programs, not the emulator itself.

File: src/gui/sdlmain.cpp [hunk 3]
```diff
@@ -4118,6 +4118,80 @@ static std::vector<std::string> get_sdl_texture_renderers()
...
+	        "                           svga_paradise, vesa_nolfb, vesa_oldvbe.\n"
+	        "\n"
+	        "  -c <command>             Run the specified DOS command before running FILE.\n"
+	        "                           Multiple commands can be specified.\n"
+	        "\n"
+	        "  --noautoexec             Don't execute DOS commands from any [autoexec] sections.\n"
+	        "\n"
+	        "  --exit                   Exit after the DOS program specified by FILE has ended.\n"
+	        "                           If no FILE has been specified, execute [autoexec] and\n"
+	        "                           terminate.\n"
```
Comment (johnnovak) [inline src/gui/sdlmain.cpp:4167]: I think this needs rewording while you're at it. If we change FILE to PATH, we should explain the two different behaviours in separate sentences (when PATH is a host FS path or a disk image, or when PATH points to a DOS executable on the host FS).

File: src/gui/sdlmain.cpp [hunk 4]
```diff
@@ -4118,6 +4118,80 @@ static std::vector<std::string> get_sdl_texture_renderers()
...
+	        "\n"
+	        "  --securemode             Enable secure mode: disables the MOUNT and IMGMOUNT commands.\n"
+	        "\n"
+	        "  --socket <num>           Run nullmodem on the specified socket number.\n"
+	        "\n"
+	        "  -h, --help               Print this help message and exit.\n"
+	        "\n"
+	        "  -v, --version            Print version information and exit.\n"
+	        "\n"
+	        "FILE                       A mountable path such as a directory, drive image\n"
```
Comment (johnnovak) [inline src/gui/sdlmain.cpp:4181]: Isn't it customary to put the description of such FILE / PATH / DIR arguments at the top of the help text, before the detailing the various options?

File: src/gui/sdlmain.cpp [hunk 5]
```diff
@@ -4170,9 +4170,7 @@ static void messages_add_sdl()
 	        "\n"
 	        "  --erasemapper            Delete the default mapper file.\n"
 	        "\n"
-	        "  --securemode             Enable secure mode. The [config] and [autoexec] sections\n"
-	        "                           of the loaded configurations will be ignored, and\n"
-	        "                           commands such as MOUNT and IMGMOUNT are disabled.\n"
+	        "  --securemode             Enable secure mode: disables the MOUNT and IMGMOUNT commands.\n"
```
Comment (johnnovak) [inline src/gui/sdlmain.cpp:4173]: Stylistic suggestion:

"Enable secure mode; this disables..."

At least "enables" should be "enable"; I've normalised all descriptions to use imperative mood which IMO is the preferred succinct way to describe options, settings, etc.

**Comment by Kappa971** [comment]
I would also like to point out that `dosbox /?` is not a valid command, someone on Windows might use it. 
But I imagine this could cause some problems on other operating systems...

File: src/gui/sdlmain.cpp (hunk 5 above)
Comment (kcgen) [reply src/gui/sdlmain.cpp]: Now:

```c++
"  --securemode    Enable secure mode by disabling the MOUNT and IMGMOUNT\n"
"                  commands.\n"
```

File: src/gui/sdlmain.cpp (hunk 4 above)
Comment (kcgen) [reply src/gui/sdlmain.cpp]: yup; moved to the top.

File: src/gui/sdlmain.cpp (hunk 3 above)
Comment (kcgen) [reply src/gui/sdlmain.cpp]: Simplified (and verified):

```c++
"  --exit   Exit after running '-c <command>'s and [autoexec] sections.\n"

**Comment by kcgen** [comment]
Updated PATH description:

```c++
"Usage: %s [OPTION]... [PATH]\n"
"\n"
"PATH    If PATH is a directory, it's mounted to C:.\n"
"        If PATH is a bootable disk image (IMA/IMG), it's booted.\n"
"        If PATH is a CDROM image (CUE/ISO), it's mounted to D:.\n"
"        If PATH is a DOS executable (BAT/COM/EXE), it's parent\h"
"        path is mounted to C: and the executable is run. When\n"
"        the executable exits, DOSBox Staging quits.\n"
"\n"
```

**Comment by kcgen** [comment]
> I would also like to point out that `dosbox /?` is not a valid command, someone on Windows might use it. But I imagine this could cause some problems on other operating systems...

Yes, on POSIX operating systems, '?' is a valid filename or directory (and we do handle filenames and paths on the CLI).

But an easier alternative is `-?`, which is now in the PR.
So we'll have: `--help`, `-h`, and `-?`, which even matches the order of help flags processed by Microsoft's terminal CLI.

**Comment by johnnovak** [comment]
> Updated PATH description:
> 
> ```c++
> "Usage: %s [OPTION]... [PATH]\n"
> "\n"
> "PATH    If PATH is a directory, it's mounted to C:.\n"
> "        If PATH is a bootable disk image (IMA/IMG), it's booted.\n"
> "        If PATH is a CDROM image (CUE/ISO), it's mounted to D:.\n"
> "        If PATH is a DOS executable (BAT/COM/EXE), it's parent\h"
> "        path is mounted to C: and the executable is run. When\n"
> "        the executable exits, DOSBox Staging quits.\n"
> "\n"
> ```

Looks much better @kcgen. Would it be better to write "it's mounted as the C drive." instead? "C:." at the end of the sentence is a bit strange perhaps.

**Comment by FeralChild64** [comment]
Just don't forget to synchronize help text changes with man page.

**Review by johnnovak — APPROVED** [review]
Approving @kcgen , just minor comments 😄

**Comment by Kappa971** [comment]
One last question, are there any guidelines regarding the number of characters for each line?
I imagine you can use more than 80 characters but is there a limit?

**Comment by Grounded0** [comment]
I designed man page to be a simple copy-paste job in section `OPTIONS` to reduce maintenance burden. Only line that differs is:

`--help                   Print help message and exit.`

**Comment by johnnovak** [comment]
> One last question, are there any guidelines regarding the number of characters for each line? I imagine you can use more than 80 characters but is there a limit?

Longest line is 86 chars currently, so best to keep it below 90.

There's no "hard limit" we need to adhere to, but I'd advise using some sensible limit anyway due to people running terminal multiplexers such as tmux.

More importantly, the recommended average line length for the best readability is 50-60 characters.
https://baymard.com/blog/line-length-readability

As it turns out, the actual indented option descriptions are currently _right in the middle_ of that sweet spot—nice! So let's keep it that way.

**Comment by kcgen** [comment]
I'm not making good progress on the man page. Maybe this can be done as a follow up PR.

@Grounded0, would greatly appreciate your help, if you can touch up the new PATH, `--autoexec`, `--exit`, and `--securemode` text.

**Comment by Grounded0** [comment]
I'll fix up that nroff. Its a messed up language.

**Comment by Kappa971** [comment]
There is a problem displaying extended characters, both in Windows Powershell and in CMD:

immagine
immagine

**Comment by Grounded0** [comment]
@Kappa971

Needs to be in UTF-16? 😁

**Comment by Grounded0** [comment]
Windows is natively UTF-16 so grabbing my 🍿.

**Comment by Kappa971** [comment]
Windows CMD uses CP850, while Powershell should be UTF-8 but apparently this is not the case. Maybe it uses CP1252? In any case, everything is at the default Windows settings (Italian).
I suspect it might be harder than expected...

**Comment by Grounded0** [comment]
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_character_encoding?view=powershell-7.3

> - utf8NoBOM: Encodes in UTF-8 format without Byte Order Mark (BOM)

> PowerShell defaults to utf8NoBOM for all output.

**Comment by Kappa971** [comment]
> https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_character_encoding?view=powershell-7.3
> 
> > * utf8NoBOM: Encodes in UTF-8 format without Byte Order Mark (BOM)
> 
> > PowerShell defaults to utf8NoBOM for all output.

So what is happening here?

**Comment by Grounded0** [comment]
Maybe you didn't save those files as UTF-8? Accidents do happen.

**Comment by FeralChild64** [comment]
https://learn.microsoft.com/en-us/windows/console/getconsoleoutputcp

Unfortunately, not all these code pages are supported by our Unicode engine…

**Comment by Grounded0** [comment]
To put it simply PowerShell may support UTF-8 natively but Windows API is still UTF-16.

**Comment by FeralChild64** [comment]
I have compared Windows code pages specified on page https://learn.microsoft.com/en-us/windows/win32/intl/code-page-identifiers with our code page set specified in the file https://github.com/dosbox-staging/dosbox-staging/blob/main/contrib/resources/mapping/MAIN.TXT - good news: there seems to be no number conflict :D

We might need to add iso-8859 series encodings (Windows code pages 28591 for ISO 8859-1, 28592 for ISO 8859-2, etc.) - boring job, but doable for 0.81.0.

Indeed, we do not support UTF-16 - but adding it shouldn't be problematic, if needed. In fact, it will probably be needed one day nevertheless, I know at least long file names on FAT32 are using UTF-16 encoding.

---

I just need a confirmation what is needed from the Unicode engine.

**Comment by Kappa971** [comment]
So currently the DOSBox help screen can't be displayed correctly for non-English languages because DOSBox doesn't support UTF-16?

**Comment by FeralChild64** [comment]
@Kappa971 It’s probably a little bit more complicated. IMHO Windows needs a little more code to determine which encoding the current console uses.

For UTF-8 console nothing has to be done besides the current processing.
If the console uses some code page, DOSBox has to use internal Unicode engine to convert the UTF-8 help text to this particular code page. It’s possible the engine will have to be extended to support few more code pages.
If console is UTF-16 (I highly doubt this is the case), than I have to add UTF-16 encoding support to our Unicode engine.

**Comment by Kappa971** [comment]
> @Kappa971 It’s probably a little bit more complicated. IMHO Windows needs a little more code to determine which encoding the current console uses.
> 
> For UTF-8 console nothing has to be done besides the current processing. If the console uses some code page, DOSBox has to use internal Unicode engine to convert the UTF-8 help text to this particular code page. It’s possible the engine will have to be extended to support few more code pages. If console is UTF-16 (I highly doubt this is the case), than I have to add UTF-16 encoding support to our Unicode engine.

Windows CMD uses the code page 850.
I don't know what encoding WIndows Powershell uses.

The code page 850 is already present in DOSBox but, apparently, something is missing or doesn't work.

**Comment by johnnovak** [comment]
FYI, the Windows code page problem with `CMD.EXE` (not PowerShell) is a long-known issue with Nim. Basically, the workaround there is to set the desired codepage manually, then output strings using the correct code page.

**Comment by Grounded0** [comment]
@kcgen 
@johnnovak 

Here's the man page ready for deployment with the requested changes:

https://anttipeltola.eu/docs/dosbox.1

**Comment by johnnovak** [comment]
> @kcgen @johnnovak
> 
> Here's the man page ready for deployment with the requested changes:
> 
> https://anttipeltola.eu/docs/dosbox.1
> 
> If you want something changed just @ me here.

Is this part of the output? 😄 

image

**Comment by Grounded0** [comment]
Its just a magic number to trigger nroff mode in Emacs. 🙂

**Comment by kcgen** [comment]
Was offline; that's a top notch man page, @Grounded0 !



Will get this added to the PR.

**Comment by kcgen** [comment]
Thankfully Microsoft's employees working on their Terminal client recognized the need for UTF-8, as it's a requirement to support emojis:



2019? We're almost half into the next decade: "Still working on it, boss!"  :motor_scooter:  :older_adult: :older_man: .. good grief.

On a more practical note - looks like you can just force UTF-8 in cmd.exe and some other with:

1. Run `intl.cpl` (which opens the regional settings in Control Panel)
2. Follow the instructions in the screen shot:
    YGaG1

From: https://stackoverflow.com/a/57134096

**Comment by kcgen** [comment]
Ok - merging this. `--help` text can now be translated: it's an incremental improvement we want, regardless of Microsoft's code-page limitations in their Terminal client.

(Another nice terminal emulator on Windows is ConEmu:  https://conemu.github.io/en/UnicodeSupport.html)

Thanks for the reviews everyone.

Thanks for tips on switching to the UTF-8 code-page, @johnnovak.

Thanks for taking up the pursuit of these Windows-specific CP's, @FeralChild64 !

**Comment by johnnovak** [comment]
> Ok - merging this. `--help` text can now be translated: it's an incremental improvement we want, regardless of Microsoft's code-page limitations in their Terminal client.
> 
> (Another nice terminal emulator on Windows is ConEmu: https://conemu.github.io/en/UnicodeSupport.html)
> 
> Thanks for the reviews everyone.
> 
> Thanks for tips on switching to the UTF-8 code-page, @johnnovak.
> 
> Thanks for taking up the pursuit of these Windows-specific CP's, @FeralChild64 !

Great executive decision @kcgen – we should not let minor details spiral out of hand and derail us, in general 👍🏻 🎖️

**Comment by Kappa971** [comment]
Thanks @kcgen!

I noticed two errors
`  --noprimaryconf          Don't load or create the the primary configuration file.`
Double "the"

`  -c <command>             Run the specified DOS command before running FILE.`
Shouldn't it be PATH?

**Comment by Grounded0** [comment]
@Kappa971 open a new ticket for these. I'm currently fixing the man page for these.

**Comment by FeralChild64** [comment]
And I’m experimenting with `GetConsoleCP` call - branch https://github.com/dosbox-staging/dosbox-staging/tree/fc/windows-help-1

**Comment by Kappa971** [comment]
> @Kappa971 open a new ticket for these. I'm currently fixing the man page for these.

Done.

**Comment by Kappa971** [comment]
> And I’m experimenting with `GetConsoleCP` call - branch https://github.com/dosbox-staging/dosbox-staging/tree/fc/windows-help-1

Tell me when I can test it 🙂

**Comment by FeralChild64** [comment]
@Kappa971 Now. PR https://github.com/dosbox-staging/dosbox-staging/pull/3064 - it will be out of draft status as soon as I fill-in the description.

**Pull request merged**